### PR TITLE
HITL - Mouse-based UI.

### PR DESCRIPTION
**Pull request by 0mdc**
## Motivation and Context

This adds a new UI to the `rearrange_v2` app with the following features:
* Mouse-based pick and place.
* Mouse-based open/close of articulated object links.
* Color coding of goal object-receptacle pairs.
* Refactoring.

---

## Features

Here are videos showing the overall changes. Note that the instruction text was disabled for visibility, and that highlight circle billboarding is only implemented in Unity.

**Double-click to open individual articulated object links:**

https://github.com/facebookresearch/habitat-lab/assets/110583667/2abf1b16-1182-432c-bcae-90ffe83f37c3

**Double-click to pick objects, and release right click to place it:**

https://github.com/facebookresearch/habitat-lab/assets/110583667/5c08ee09-ae9d-4b5a-9631-0ba9ac69b7d2

**While holding right click, an indicator shows up communicating whether the object can be placed:**

https://github.com/facebookresearch/habitat-lab/assets/110583667/1a891cb5-d43e-421b-af61-f2a19c5e2359

**Experiment objects must be placed on specific receptacles. The pairs are color-coded:**

https://github.com/facebookresearch/habitat-lab/assets/110583667/be07b754-7834-4f65-b890-d161ffb7b1ca

**When an action is possible, the color is green. Otherwise, it is red:**

https://github.com/facebookresearch/habitat-lab/assets/110583667/d4024df8-65fd-46ff-97d2-79cf423a6545

**All of this works in Unity:**

https://github.com/facebookresearch/habitat-lab/assets/110583667/c38b6a62-aad9-4134-a0aa-1a6276069856

---

## How Has This Been Tested

Tested locally and with remote Unity client.

## Types of changes

- **\[Development\]**

## Checklist

- [x] My code follows the code style of this project.
- [x] I have updated the documentation if required.
- [x] I have read the [**CONTRIBUTING**](/CONTRIBUTING.md) document.
- [x] I have completed my CLA (see **CONTRIBUTING**)
- [x] I have added tests to cover my changes if required.

### CODE REVIEW

File: examples/hitl/rearrange_v2/rearrange_v2.py [hunk 1]
```diff
@@ -244,132 +152,84 @@ def on_environment_reset(self, episode_recorder_dict):
...
+        current_episode = self._app_service.env.current_episode
+        if current_episode.info.get("extra_info") is not None:
+            extra_info = current_episode.info["extra_info"]
+            self._task_instruction = extra_info["instruction"]
+            for proposition in extra_info["evaluation_propositions"]:
+                object_ids: List[int] = []
+                object_handles = proposition["args"]["object_handles"]
+                for object_handle in object_handles:
+                    obj = sim_utilities.get_obj_from_handle(
+                        self.get_sim(), object_handle
```
Comment (jturner65) [inline examples/hitl/rearrange_v2/rearrange_v2.py:173]: perhaps cache self.get_sim() before the for loop?

File: examples/hitl/rearrange_v2/ui.py [hunk 2]
```diff
@@ -0,0 +1,448 @@
...
+                    link_index = self._get_link_index(receptacle_id)
+                    if link_index is not None:
+                        link_node = managed_object.get_link_scene_node(
+                            link_index
+                        )
+                        aabb = link_node.cumulative_bb
+                if aabb is not None:
+                    self._draw_aabb(
+                        aabb, managed_object.transformation, goal_pair_color
+                    )
```
Comment (jturner65) [inline examples/hitl/rearrange_v2/ui.py]: Here's a suggestion to limit the map/dict lookups inherent in dot-dereferencing (such as with _self_ ) in python by making local references before the loop starts to both member variables and methods that are called often within the nested loops.  

Not sure how large the object_receptacle_pairs list is, nor the subsequent sublists of rigid_ids and receptacle_ids, but if these are expected to possibly be large, these changes might make a noticeable difference in draw speed.

-shrug- ymmv :) 

```suggestion
        sim = self._sim
        obj_receptacle_pairs = self._object_receptacle_pairs
        link_id_to_ao_map = self._link_id_to_ao_map
        dest_mask = self._dest_mask
        # can do this with methods as well - note the absence of parenthesis
        get_obj_from_id = sim_utilities.get_obj_from_id
        get_link_index = self._get_link_index
        draw_gui_circle = self._gui_drawer.draw_circle
        draw_gui_aabb = self._draw_aabb
        
        for i in range(len(obj_receptacle_pairs)):
            rigid_ids = obj_receptacle_pairs[i][0]
            receptacle_ids = obj_receptacle_pairs[i][1]
            goal_pair_color = COLOR_GOALS[i % len(COLOR_GOALS)]
            for rigid_id in rigid_ids:
                if self._hover_selection.object_id == rigid_id:
                    continue
                managed_object = get_obj_from_id(
                    sim, rigid_id, link_id_to_ao_map
                )
                translation = managed_object.translation
                draw_gui_circle(
                    translation=translation,
                    radius=0.25,
                    color=goal_pair_color,
                    billboard=True,
                    destination_mask=dest_mask,
                )
            for receptacle_id in receptacle_ids:
                managed_object = get_obj_from_id(
                    sim, receptacle_id, link_id_to_ao_map
                )
                aabb = None
                if hasattr(managed_object, "collision_shape_aabb"):
                    aabb = managed_object.collision_shape_aabb
                else:
                    link_index = get_link_index(receptacle_id)
                    if link_index is not None:
                        link_node = managed_object.get_link_scene_node(
                            link_index
                        )
                        aabb = link_node.cumulative_bb
                if aabb is not None:
                    draw_gui_aabb(
                        aabb, managed_object.transformation, goal_pair_color
                    )
```

**Review by jturner65 — APPROVED** [review]
LGTM
I made a few suggestions regarding possible performance improvements, but not sure how pertinent they are.

**Comment by 0mdc** [comment]
Thanks John! I applied your suggestions.

**Pull request merged**